demo/parser_demo.cpp

```diff
@@ -5,140 +5,216 @@
+        {
             specific->accept(*this);
-        } else if (ast::AliasedType *specific = dynamic_cast<ast::AliasedType*>(node); specific != nullptr) {
+        }
+        else if (ast::AliasedType *specific = dynamic_cast<ast::AliasedType *>(node); specific != nullptr)
+        {
             fmt::print("{:|>{}}- [Type Identifier]> {}\n", "", depth, specific->name.lit);
-        } else {
+        }
+        else
+        {
             fmt::print("{:|>{}}- [Type]> (unknown type)\n", "", depth);
         }
     };
-    void visit(ast::PrimitiveType *node) override {
-        if (node == nullptr) {
+    void visit(ast::PrimitiveType *node) override
+    {
+        if (node == nullptr)
+        {
             fmt::print("{:|>{}}- [PrimitiveType]> null\n", "", depth);
-        } else {
+        }
+        else
+        {
             fmt::print("{:|>{}}- [PrimitiveType]> {}\n", "", depth, node->type.lit);
         }
     };
-    void visit(ast::ArrayType *node) override {
-        if (node == nullptr) {
+    void visit(ast::ArrayType *node) override
+    {
+        if (node == nullptr)
+        {
             fmt::print("{:|>{}}- [ArrayType]> null\n", "", depth);
-        } else {
+        }
+        else
+        {
             fmt::print("{:|>{}}- [ArrayType]>\n", "", depth);
             depth++;
             node->length->accept(*this);
             node->elementType->accept(*this);
             depth--;
         }
     };
-    void visit(ast::RecordType *node) override {
-        if (node == nullptr) {
+    void visit(ast::RecordType *node) override
+    {
+        if (node == nullptr)
+        {
             fmt::print("{:|>{}}- [RecordType]> null\n", "", depth);
-        } else {
+        }
+        else
+        {
             fmt::print("{:|>{}}- [RecordType]>\n", "", depth);
             depth++;
-            for (auto field : node->fields) {
+            for (auto field : node->fields)
+            {
                 field->accept(*this);
             }
             depth--;
         }
     };
-    void visit(ast::Variable *node) override {
+    void visit(ast::Variable *node) override{
 
     };
-    void visit(ast::Body *node) override {
+    void visit(ast::Body *node) override{
 
     };
-    void visit(ast::Statement *node) override {
+    void visit(ast::Statement *node) override{
 
     };
-    void visit(ast::Assignment *node) override {
+    void visit(ast::Assignment *node) override{
 
     };
-    void visit(ast::RoutineCall *node) override {
+    void visit(ast::RoutineCall *node) override{
 
     };
-    void visit(ast::WhileLoop *node) override {
-
+    void visit(ast::WhileLoop *node) override
+    {
+        if (node == nullptr)
+        {
+            fmt::print("{:|>{}}- [WhileLoop]> null\n", "", depth);
+        }
+        else
+        {
+            fmt::print("{:|>{}}- [WhileLoop]>\n", "", depth);
+            depth++;
+            node->condition->accept(*this);
+            node->body->accept(*this);
```

Comment:
This is a little bit tricky since `body` can technically be null (if there is no body), I guess. Or would it be an `ast::Body` object without any meaningful content?

Reply:
It is impossible to check it my stage. Also I'm not sure that our grammar gives us a possibility to have empty `Body`

Reply:
A body can always be empty. It is defined as `Body ::= { SimpleDeclaration | Statement }`

Reply:
> A body can always be empty. It is defined as `Body ::= { SimpleDeclaration | Statement }`

Yeah, but then there should be a body node with zero decls and statements....